bootloader/grub.go

```diff
@@ -180,27 +191,84 @@ func (g *grub) RemoveKernelAssets(s snap.PlaceInfo) error {
 func (g *grub) makeKernelEfiSymlink(s snap.PlaceInfo, name string) error {
 	// use a relative symlink destination so that it resolves properly, if grub
 	// is located at /run/mnt/ubuntu-boot or /boot/grub, etc.
-	extractedKernel := filepath.Join(
+	target := filepath.Join(
 		filepath.Base(s.MountFile()),
 		"kernel.efi",
 	)
 
+	// the location of the destination symlink as an absolute filepath
+	source := filepath.Join(g.dir(), name)
+
 	// check that the kernel snap has been extracted already so we don't
 	// inadvertently create a dangling symlink
 	// expand the relative symlink from g.dir()
-	if !osutil.FileExists(filepath.Join(g.dir(), extractedKernel)) {
+	if !osutil.FileExists(filepath.Join(g.dir(), target)) {
 		return fmt.Errorf(
 			"cannot enable %s at %s: %v",
 			name,
-			extractedKernel,
+			target,
 			os.ErrNotExist,
 		)
 	}
 
-	return os.Symlink(
-		extractedKernel,
-		filepath.Join(g.dir(), name),
-	)
+	// check if the target symlink already exists, if it doesn't then we can
+	// just create it directly
+	// if it doesn't we need to be more tricky
+	_, err := os.Lstat(source)
+	if err == nil {
+		// symlink exists
+
+		// we want to create the kernel.efi symlink atomically, so that we don't
+		// end up in a situation where kernel.efi is not present on the filesystem
+		// so we create an intermediate symlink which points to the kernel, then
+		// swap that with the real kernel.efi symlink and delete the intermediate
+		// symlink
+		// This ensures that we never end up in a state without a kernel.efi and
+		// also that we don't need a try-kernel.efi to pre-exist (on initial install
+		// it will not, so we can't just swap try-kernel.efi and kernel.efi)
+
+		// the intermediate symlink - use a random name to be minimize chance that
+		// the symlink already exists
+		intermediate := source + "." + strconv.Itoa(r.Intn(100))
+
+		// still be safe and remove the symlink if it exists
+		_, err := os.Lstat(intermediate)
```

Comment:
If we go with a known `kernel-candidate.efi` symlink, we could just `os.Remove()` it and bail out if `err != nil && !os.IsNotExist(err)`